File: client/src/lib/apiClient.ts
import { 
  LoginRequest, 
  RegisterUserRequest, 
  UpdateUserRequest, 
  LogExerciseRequest,
  LoginResponse,
  UserResponse,
  ExerciseResponse,
  LeaderboardEntry 
} from './types';
import config from './config';

// REMOVE the static API_BASE_URL and replace with a function that gets the current value
// This ensures we always use the most up-to-date URL after port discovery completes
const getApiBaseUrl = (): string => config.api.baseUrl;

// Helper function to get the JWT token from storage
const getToken = (): string | null => {
  return localStorage.getItem(config.auth.storageKeys.token);
};

/**
 * Normalize the API endpoint to handle inconsistencies in path configuration
 * @param endpoint API endpoint path
 * @returns Normalized endpoint path
 */
const normalizeEndpoint = (endpoint: string): string => {
  // Try both with and without the /v1 prefix
  // This handles potential inconsistencies between port discovery and server routes
  const apiBaseUrl = getApiBaseUrl(); // Get current base URL
  const hasV1InBaseUrl = apiBaseUrl.includes('/v1');
  
  if (hasV1InBaseUrl && endpoint.startsWith('/v1/')) {
    // If base URL already has /v1 and endpoint also starts with /v1, remove from endpoint
    return endpoint.substring(3);
  }
  
  // Otherwise, just ensure endpoint starts with /
  return endpoint.startsWith('/') ? endpoint : `/${endpoint}`;
};

/**
 * Helper function for making API requests with automatic retries
 * @param endpoint API endpoint
 * @param method HTTP method
 * @param body Request body (optional)
 * @param requiresAuth Whether authentication is required
 * @param retries Number of retries to attempt
 * @returns Promise with the response data
 */
const apiRequest = async <T>(
  endpoint: string, 
  method: string, 
  body?: any, 
  requiresAuth: boolean = false,
  retries: number = 1
): Promise<T> => {
  const headers: HeadersInit = {
    'Content-Type': 'application/json',
  };

  if (requiresAuth) {
    const token = getToken();
    if (!token) {
      // Handle cases where auth is required but no token is found
      throw new Error('Authentication required.');
    }
    headers['Authorization'] = `Bearer ${token}`;
  }

  const requestConfig: RequestInit = {
    method: method,
    headers: headers,
  };

  if (body) {
    requestConfig.body = JSON.stringify(body);
  }

  const makeRequest = async (): Promise<T> => {
    try {
      // Always get the current base URL for each request
      const apiBaseUrl = getApiBaseUrl();
      const normalizedEndpoint = normalizeEndpoint(endpoint);
      const apiUrl = `${apiBaseUrl}${normalizedEndpoint}`;
      
      console.log(`Making request to: ${apiUrl}`, { 
        method, 
        requiresAuth,
        hasToken: !!getToken()
      });
      
      const response = await fetch(apiUrl, requestConfig);

      // Check if the response is ok (status in the range 200-299)
      if (!response.ok) {
        let errorData;
        try {
          // Try to parse the error response body
          errorData = await response.json();
        } catch (e) {
          // If parsing fails, use the status text
          errorData = { error: response.statusText };
        }
        console.error(`API Error (${response.status}):`, errorData);
        // Throw an error with details from the backend if possible
        throw new Error(errorData.error || `HTTP error! status: ${response.status}`);
      }

      // If response is OK, try to parse JSON, handle empty responses (e.g., 204 No Content)
      const contentType = response.headers.get('content-type');
      if (contentType && contentType.indexOf('application/json') !== -1) {
        if (response.status === 204) {
          return null as T; // Handle No Content response
        }
        return await response.json();
      } else {
        // Handle non-JSON responses if necessary
        return null as T;
      }
    } catch (error) {
      // Log the error
      console.error(`API request failed: ${method} ${endpoint}`, error);
      // Rethrow to allow caller to handle
      throw error;
    }
  };

  // Try the request with retries
  let lastError: Error | null = null;
  for (let attempt = 0; attempt <= retries; attempt++) {
    try {
      return await makeRequest();
    } catch (error) {
      lastError = error instanceof Error ? error : new Error(String(error));
      
      // If this was our last retry, or it's an error we shouldn't retry (like auth errors)
      if (
        attempt === retries || 
        (error instanceof Error && error.message === 'Authentication required.')
      ) {
        throw lastError;
      }
      
      // Wait before retry with exponential backoff
      const delay = Math.pow(2, attempt) * 100; // 100ms, 200ms, 400ms, etc.
      console.log(`Retry ${attempt + 1}/${retries} after ${delay}ms`);
      await new Promise(resolve => setTimeout(resolve, delay));
    }
  }
  
  // This should never be reached due to the throw in the loop
  throw lastError || new Error('Unknown API request error');
};

// --- Auth Endpoints ---

export const registerUser = (data: RegisterUserRequest): Promise<UserResponse> => {
  return apiRequest<UserResponse>('/users/register', 'POST', data, false, 2);
};

export const loginUser = (data: LoginRequest): Promise<LoginResponse> => {
  return apiRequest<LoginResponse>('/users/login', 'POST', data, false, 2);
};

// --- User Endpoints ---

export const updateCurrentUser = (data: UpdateUserRequest): Promise<UserResponse> => {
  return apiRequest<UserResponse>('/users/me', 'PATCH', data, true, 1);
};

// --- Exercise Endpoints ---

export const logExercise = (data: LogExerciseRequest): Promise<ExerciseResponse> => {
  return apiRequest<ExerciseResponse>('/exercises', 'POST', data, true, 1);
};

export const getUserExercises = (): Promise<ExerciseResponse[]> => {
  // Try both the standard and alternative endpoints
  return apiRequest<ExerciseResponse[]>('/user-exercises', 'GET', null, true, 1)
    .catch(error => {
      console.log('Failed with /user-exercises, trying fallback endpoint...');
      // Use the fallback endpoint that exists in routes.ts
      return apiRequest<ExerciseResponse[]>('/exercises/user', 'GET', null, true, 1);
    });
};

// --- Leaderboard Endpoints ---

export const getLeaderboard = (exerciseType: string): Promise<LeaderboardEntry[]> => {
  return apiRequest<LeaderboardEntry[]>(`/leaderboard/${exerciseType}`, 'GET', null, false, 1);
};

// --- Helper functions for auth state ---

export const storeAuthData = (token: string, user: UserResponse): void => {
  localStorage.setItem(config.auth.storageKeys.token, token);
  localStorage.setItem(config.auth.storageKeys.user, JSON.stringify(user));
};

export const clearAuthData = (): void => {
  localStorage.removeItem(config.auth.storageKeys.token);
  localStorage.removeItem(config.auth.storageKeys.user);
};

export const getStoredUser = (): UserResponse | null => {
  const userData = localStorage.getItem(config.auth.storageKeys.user);
  try {
    return userData ? JSON.parse(userData) : null;
  } catch (error) {
    console.error('Failed to parse stored user data:', error);
    return null;
  }
}; 

File: PTChampion-Swift/Data/Network/APIClient.swift
import Foundation
// Removed Combine import
import PTChampion_Swift // Assuming models are accessible via the main module

// MARK: - API Client
class APIClient {
    static let shared = APIClient()

    private let baseURL = URL(string: "http://localhost:5000/api")! // Standardized base URL
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder
    private let authTokenManager = AuthTokenManager.shared // Placeholder for token management

    // Current user cache - consider moving to a dedicated UserSession manager
    private var currentUser: User?

    private init() {
        let configuration = URLSessionConfiguration.default
        self.session = URLSession(configuration: configuration)

        self.decoder = JSONDecoder()
        self.decoder.keyDecodingStrategy = .convertFromSnakeCase
        self.decoder.dateDecodingStrategy = .iso8601

        self.encoder = JSONEncoder()
        self.encoder.keyEncodingStrategy = .convertToSnakeCase
        self.encoder.dateEncodingStrategy = .iso8601
    }

    // MARK: - Token Management (Delegated)

    // Placeholder functions demonstrating delegation
    private func getToken() -> String? {
        return authTokenManager.getToken()
    }

    private func saveAuthData(token: String, expiresIn: String?, user: User) {
         authTokenManager.saveToken(token: token, expiresIn: expiresIn)
         // Consider saving user ID or full user object via AuthTokenManager or separate UserSession
         // authTokenManager.saveUserID(user.id)
         self.currentUser = user // Keep local cache for now
    }

    private func clearAuthData() {
        authTokenManager.clearToken()
        // authTokenManager.clearUserID()
        self.currentUser = nil
    }


    // MARK: - Health Check

    func checkApiHealth() async throws -> HealthResponse {
        return try await makeRequest(endpoint: "/health", method: "GET", requiresAuth: false)
    }

    // MARK: - Authentication

    func login(username: String, password: String) async throws -> User {
        let loginData = ["username": username, "password": password]
        let authResponse: AuthResponse = try await makeRequest(
            endpoint: "/login",
            method: "POST",
            body: loginData,
            requiresAuth: false,
            additionalHeaders: ["X-Client-Platform": "mobile"]
        )

        // Save token and user
        saveAuthData(token: authResponse.token, expiresIn: authResponse.expiresIn, user: authResponse.user)
        return authResponse.user
    }

    func register(username: String, password: String) async throws -> User {
        let registerData = ["username": username, "password": password]
         let authResponse: AuthResponse = try await makeRequest(
             endpoint: "/register",
             method: "POST",
             body: registerData,
             requiresAuth: false,
             additionalHeaders: ["X-Client-Platform": "mobile"]
         )

        // Save token and user
        saveAuthData(token: authResponse.token, expiresIn: authResponse.expiresIn, user: authResponse.user)
        return authResponse.user
    }

    func validateToken() async throws -> Bool {
        guard getToken() != nil else {
            return false
        }

        do {
            let _: User = try await makeRequest(endpoint: "/validate-token", method: "GET", requiresAuth: true)
            return true
        } catch let error as NetworkError where error == .unauthorized {
            clearAuthData()
            return false
        } catch {
            // Rethrow other errors
            throw error
        }
    }

    func logout() async throws {
         clearAuthData()
         // Optional: Call server-side logout endpoint, ignoring errors
         Task { // Fire and forget
             try? await makeRequest(endpoint: "/logout", method: "POST", requiresAuth: false, responseType: EmptyResponse.self)
         }
    }

    func getCurrentUser() async throws -> User {
        // If we have a cached user, return it
        if let user = currentUser {
            return user
        }

        // Otherwise fetch from API
        let user: User = try await makeRequest(endpoint: "/user", method: "GET", requiresAuth: true)
        self.currentUser = user // Update cache
        return user
    }

    // MARK: - User Location

    func updateUserLocation(latitude: Double, longitude: Double) async throws -> User {
        let locationData = ["latitude": latitude, "longitude": longitude]
        let user: User = try await makeRequest(endpoint: "/user/location", method: "POST", body: locationData, requiresAuth: true)
        self.currentUser = user // Update cache
        return user
    }

    // MARK: - Exercises (Public endpoints)

    func getExercises() async throws -> [Exercise] {
        return try await makeRequest(endpoint: "/exercises", method: "GET", requiresAuth: false)
    }

    func getExerciseById(id: Int) async throws -> Exercise {
        return try await makeRequest(endpoint: "/exercises/\\(id)", method: "GET", requiresAuth: false)
    }

    // MARK: - User Exercises (Protected endpoints)

    func getUserExercises() async throws -> [UserExercise] {
        return try await makeRequest(endpoint: "/user-exercises", method: "GET", requiresAuth: true)
    }

    func getUserExercisesByType(type: String) async throws -> [UserExercise] {
        return try await makeRequest(endpoint: "/user-exercises/\\(type)", method: "GET", requiresAuth: true)
    }

    func getLatestUserExercises() async throws -> [String: UserExercise] {
         return try await makeRequest(endpoint: "/user-exercises/latest/all", method: "GET", requiresAuth: true)
    }

    func createUserExercise(userExercise: CreateUserExerciseRequest) async throws -> UserExercise {
        return try await makeRequest(endpoint: "/user-exercises", method: "POST", body: userExercise, requiresAuth: true)
    }

    // MARK: - Leaderboard (Public endpoints)

    func getGlobalLeaderboard() async throws -> [LeaderboardEntry] {
        // Assuming LeaderboardEntry is defined in Data/Models/Leaderboard.swift
        // Might need adjustments based on actual backend response structure
        return try await makeRequest(endpoint: "/leaderboard/global", method: "GET", requiresAuth: false)
    }

    func getLocalLeaderboard(latitude: Double, longitude: Double, radiusMiles: Int = 5) async throws -> [LeaderboardEntry] {
        // Assuming LeaderboardEntry is defined in Data/Models/Leaderboard.swift
        let endpoint = "/leaderboard/local?latitude=\\(latitude)&longitude=\\(longitude)&radius=\\(radiusMiles)"
        return try await makeRequest(endpoint: endpoint, method: "GET", requiresAuth: false)
    }

    // MARK: - Sync Operations

    func syncUserData(deviceId: String) async throws -> SyncResponse {
        // Get the last sync timestamp, or use a default if this is the first sync
        // This logic should likely move to a dedicated SyncManager or Repository
        let lastSyncTimestamp = await getLastSyncTimestamp() // Placeholder for repository/sync manager call

        // Get any pending (unsynced) user exercises
        // This logic should live in a local data source/repository
        let unsyncedExercises = await getUnsyncedExercisesFromLocal() // Placeholder

        // Create the sync request
        let syncRequest = SyncRequest(
            deviceId: deviceId,
            lastSyncTimestamp: lastSyncTimestamp,
            data: SyncData(
                userExercises: unsyncedExercises,
                profile: nil // Profile updates might be handled separately
            )
        )

        let response: SyncResponse = try await makeRequest(endpoint: "/sync", method: "POST", body: syncRequest, requiresAuth: true)

        // Handle response: Save new timestamp, update local cache
        // This post-sync processing should also move to SyncManager/Repository
        await saveNewSyncTimestamp(response.timestamp) // Placeholder for repository/sync manager call

        if let exercises = response.data?.userExercises {
             await saveExercisesToLocalCache(exercises) // Placeholder
        }
        if let profile = response.data?.profile {
            self.currentUser = profile // Update local cache
             // Potentially notify AuthTokenManager or UserSession
        }

        return response
    }

    // Placeholder for fetching last sync timestamp - move to a repository/data source
    private func getLastSyncTimestamp() async -> String {
        // Implementation would fetch from a sync status store (UserDefaults, CoreData, etc.)
        // Managed by a dedicated sync component
        return "2000-01-01T00:00:00.000Z" // Default value
    }

    // Placeholder for saving new sync timestamp - move to a repository/data source
    private func saveNewSyncTimestamp(_ timestamp: String) async {
         // Implementation would save to a sync status store
    }

    // Placeholder for fetching local data - move to a repository/data source
    private func getUnsyncedExercisesFromLocal() async -> [CreateUserExerciseRequest] {
        // Implementation would fetch from CoreData or other local storage
        return []
    }

    // Placeholder for saving synced data - move to a repository/data source
    private func saveExercisesToLocalCache(_ exercises: [UserExercise]) async {
        // Implementation would save/update CoreData or other local storage
    }

    // MARK: - Profile Update
    func updateProfile(profileData: UpdateProfileRequest) async throws -> User {
         let user: User = try await makeRequest(endpoint: "/profile", method: "POST", body: profileData, requiresAuth: true)
         self.currentUser = user // Update cache
         return user
    }


    // MARK: - Generic Request Method (async/await)

    private func makeRequest<T: Decodable, B: Encodable>(
        endpoint: String,
        method: String,
        queryItems: [URLQueryItem]? = nil,
        body: B? = nil,
        requiresAuth: Bool = false,
        additionalHeaders: [String: String] = [:],
        responseType: T.Type = T.self // Keep explicit type for clarity
    ) async throws -> T {

        guard var urlComponents = URLComponents(url: baseURL.appendingPathComponent(endpoint), resolvingAgainstBaseURL: true) else {
            throw NetworkError.badURL
        }
        urlComponents.queryItems = queryItems // Add query items if any

        guard let url = urlComponents.url else {
             throw NetworkError.badURL
         }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept") // Prefer Accept header

        // Add Auth Token if required
        if requiresAuth {
            guard let token = getToken() else {
                // No token available for authenticated request
                throw NetworkError.unauthorized // Or a more specific .noToken error if defined
            }
            request.setValue("Bearer \\(token)", forHTTPHeaderField: "Authorization")
        }

        // Add any additional headers
        additionalHeaders.forEach { key, value in
            request.setValue(value, forHTTPHeaderField: key)
        }

        // Encode body if provided
        if let body = body {
            // Only set Content-Type if there is a body
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            do {
                request.httpBody = try encoder.encode(body)
            } catch {
                // Log encoding error details
                print("Encoding error: \\(error)")
                throw NetworkError.requestFailed(error) // Or a specific encoding error type
            }
        }

        // Perform the request
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
             // Log network request error
             print("Network request error: \\(error)")
            // Map specific URLSession errors if needed (e.g., timeout, offline)
            throw NetworkError.requestFailed(error)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
             print("Invalid response: Not an HTTPURLResponse")
            throw NetworkError.invalidResponse(response)
        }

        // Log response status and potentially headers/body for debugging
        print("[\(method)] \(url.absoluteString) - Status: \(httpResponse.statusCode)")
        // print("Response Headers: \\(httpResponse.allHeaderFields)")
        // print("Response Body: \\(String(data: data, encoding: .utf8) ?? "<No Body or Non-UTF8>")")


        // Handle HTTP status codes
        switch httpResponse.statusCode {
        case 200...299:
            // Handle empty response (e.g., 204 No Content)
            if T.self == EmptyResponse.self {
                 // Ensure EmptyResponse is defined somewhere (e.g., Data/Models/Common.swift)
                 guard let empty = EmptyResponse() as? T else {
                     // This should ideally not happen if T is constrained correctly
                     throw NetworkError.decodingError(DecodingError.typeMismatch(T.self, DecodingError.Context(codingPath: [], debugDescription: "Could not cast EmptyResponse to expected type T")))
                 }
                 return empty
             }
            if data.isEmpty && !(T.self == EmptyResponse.self) {
                // Received empty data for a non-empty expected response type
                print("Warning: Received empty data for expected type \\(T.self)")
                 throw NetworkError.decodingError(DecodingError.dataCorrupted(DecodingError.Context(codingPath: [], debugDescription: "Expected non-empty data but received empty.")))
            }

            // Decode successful response
            do {
                let decodedObject = try decoder.decode(T.self, from: data)
                return decodedObject
            } catch let decodingError as DecodingError {
                 // Provide detailed decoding error logging
                 print("Decoding error: \\(decodingError)")
                 Self.logDecodingErrorDetails(decodingError, data: data)
                 throw NetworkError.decodingError(decodingError)
             } catch {
                 print("Generic decoding error: \\(error)")
                 throw NetworkError.decodingError(error)
             }
        case 401: // Unauthorized
            // Clear token, potentially trigger re-login flow via notification/delegate
            if requiresAuth { // Only clear token if the failed request required auth
                 clearAuthData()
            }
            throw NetworkError.unauthorized
        case 403: // Forbidden
            // Often indicates lack of permissions even if authenticated
            // Decode potential error body
             let apiError = try? decoder.decode(APIErrorResponse.self, from: data)
             print("Forbidden (403): \\(apiError?.message ?? "No error message")")
             throw NetworkError.apiError(apiError ?? APIErrorResponse(error: "Forbidden", message: String(data: data, encoding: .utf8), code: 403)) // Create a default if decode fails

        case 404: // Not Found
            let apiError = try? decoder.decode(APIErrorResponse.self, from: data)
             print("Not Found (404): \\(apiError?.message ?? "No error message")")
             throw NetworkError.apiError(apiError ?? APIErrorResponse(error: "Not Found", message: String(data: data, encoding: .utf8), code: 404))

        case 400, 422: // Bad Request, Unprocessable Entity (often validation errors)
             do {
                 let apiError = try decoder.decode(APIErrorResponse.self, from: data)
                 print("API Error (\(httpResponse.statusCode)): \\(apiError.message ?? apiError.error)")
                 throw NetworkError.apiError(apiError)
             } catch let decodingError {
                 // Fallback if error response itself is not decodable
                 print("Failed to decode API error response (\(httpResponse.statusCode)): \\(decodingError)")
                 print("Raw error data: \\(String(data: data, encoding: .utf8) ?? "<invalid encoding>")")
                 // Create a generic error using the status code and raw data
                  let fallbackMessage = String(data: data, encoding: .utf8) ?? "Undecodable error response"
                  throw NetworkError.apiError(APIErrorResponse(error: "Server Error", message: fallbackMessage, code: httpResponse.statusCode))
             }

        case 500...599: // Server Errors
            let apiError = try? decoder.decode(APIErrorResponse.self, from: data)
            print("Server Error (\(httpResponse.statusCode)): \\(apiError?.message ?? "No error message")")
            throw NetworkError.apiError(apiError ?? APIErrorResponse(error: "Internal Server Error", message: String(data: data, encoding: .utf8), code: httpResponse.statusCode))


        default:
             print("Unhandled HTTP status code: \(httpResponse.statusCode)")
             // Attempt to decode as APIErrorResponse, otherwise use generic invalid response
             let apiError = try? decoder.decode(APIErrorResponse.self, from: data)
             if let apiError = apiError {
                 throw NetworkError.apiError(apiError)
             } else {
                 throw NetworkError.invalidResponse(httpResponse)
             }
        }
    }

    // Overload for requests without a request body
     private func makeRequest<T: Decodable>(
         endpoint: String,
         method: String,
         queryItems: [URLQueryItem]? = nil,
         requiresAuth: Bool = false,
         additionalHeaders: [String: String] = [:],
         responseType: T.Type = T.self
     ) async throws -> T {
         // Use a placeholder Encodable type like `EmptyBody` for the body parameter
         // Ensure EmptyBody is defined (e.g., in Data/Models/Common.swift)
         struct EmptyBody: Encodable {}
         return try await makeRequest(endpoint: endpoint, method: method, queryItems: queryItems, body: EmptyBody?.none, requiresAuth: requiresAuth, additionalHeaders: additionalHeaders, responseType: responseType)
     }

     // Overload for requests without a response body (e.g., POST/PUT returning 204)
     private func makeRequest<B: Encodable>(
         endpoint: String,
         method: String,
         queryItems: [URLQueryItem]? = nil,
         body: B? = nil,
         requiresAuth: Bool = false,
         additionalHeaders: [String: String] = [:]
     ) async throws {
         // Call the generic function expecting `EmptyResponse`
         // Ensure EmptyResponse is defined (e.g., in Data/Models/Common.swift)
         let _: EmptyResponse = try await makeRequest(endpoint: endpoint, method: method, queryItems: queryItems, body: body, requiresAuth: requiresAuth, additionalHeaders: additionalHeaders, responseType: EmptyResponse.self)
     }

    // Helper for logging decoding errors
    private static func logDecodingErrorDetails(_ error: DecodingError, data: Data) {
        print("--- Decoding Error Details ---")
        switch error {
        case .typeMismatch(let type, let context):
            print("Type mismatch for type \\(type) at path: \\(context.codingPath.map { $0.stringValue }.joined(separator: "."))")
            print("Debug description: \\(context.debugDescription)")
        case .valueNotFound(let type, let context):
            print("Value not found for type \\(type) at path: \\(context.codingPath.map { $0.stringValue }.joined(separator: "."))")
            print("Debug description: \\(context.debugDescription)")
        case .keyNotFound(let key, let context):
            print("Key not found: '\\(key.stringValue)' at path: \\(context.codingPath.map { $0.stringValue }.joined(separator: "."))")
            print("Debug description: \\(context.debugDescription)")
        case .dataCorrupted(let context):
            print("Data corrupted at path: \\(context.codingPath.map { $0.stringValue }.joined(separator: "."))")
            print("Debug description: \\(context.debugDescription)")
        @unknown default:
            print("Unknown decoding error: \\(error.localizedDescription)")
        }
        if let jsonString = String(data: data, encoding: .utf8) {
             print("Raw JSON data trying to be decoded: \\(jsonString)")
        } else {
             print("Raw data trying to be decoded was not valid UTF-8.")
        }
        print("-----------------------------")
    }
}

// Removed placeholder AuthTokenManager class

// Removed placeholder EmptyResponse struct

// Note: Ensure all referenced Models (User, Exercise, UserExercise, LeaderboardEntry,
// AuthResponse, HealthResponse, CreateUserExerciseRequest, SyncRequest, SyncData,
// SyncResponse, UpdateProfileRequest, EmptyResponse, EmptyBody) and Errors (NetworkError, APIErrorResponse)
// are correctly defined and accessible. You might need to adjust imports or use `@_exported import`. 